pkg/api/api.go

```diff
@@ -263,6 +303,43 @@ func (s *server) newTracingHandler(spanName string) func(h http.Handler) http.Ha
 	}
 }
 
+func (s *server) permissionCheckHandler() func(h http.Handler) http.Handler {
+	if !s.Restricted {
+		return func(h http.Handler) http.Handler {
+			return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				h.ServeHTTP(w, r)
+			})
+		}
+	}
+
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			reqToken := r.Header.Get("Authorization")
+			if !strings.HasPrefix(reqToken, "Bearer ") {
+				http.Error(w, "Missing bearer token", http.StatusForbidden)
+				return
+			}
+
+			keys := strings.Split(reqToken, "Bearer ")
+
+			if len(keys) != 2 || strings.Trim(keys[1], " ") == "" {
+				http.Error(w, "Missing API Key", http.StatusForbidden)
+				return
+			}
+
+			apiKey := keys[1]
+
+			allowed := s.auth.Enforce(apiKey, r.URL.Path, r.Method)
+			if !allowed {
+				w.WriteHeader(http.StatusForbidden)
```

Comment:
I think that this response deserves a body, json encoded, explaining that the auth key is not allowed.